mlflow/pyfunc/model.py

```diff
@@ -73,16 +74,57 @@ def _log_warning_if_params_not_in_predict_signature(logger, params):
         )
 
 
-class PythonModel:
+class _PythonModelMetaclass(ABCMeta):
+    def __call__(cls, *args, **kwargs):
+        cls._warn_on_setting_model_in_init()
+        return super().__call__(*args, **kwargs)
+
+    def _warn_on_setting_model_in_init(cls):
+        if not cls.__bases__ == (PythonModel,):
+            return
+
+        import click
+
+        message = (
+            click.style(
+                "It looks like you're trying to save a model as an instance attribute. ",
+                fg="yellow",
+            )
+            + click.style("This is not recommended ", fg="yellow", bold=True)
+            + click.style(
+                "as it can cause problems with model serialization, "
+                "especially for large models. ",
+                fg="yellow",
+            )
+            + click.style("Please use the `artifacts` parameter", fg="yellow", bold=True)
+            + click.style(
+                ", and load your external model in the `load_context()` method instead.",
+                fg="yellow",
+            )
+        )
+
+        try:
+            model_param_present = "model" in inspect.signature(cls.__init__).parameters
+            model_assignment = (
+                cls.load_context == PythonModel.load_context
+                and "self.model =" in inspect.getsource(cls.__init__)
+            )
+            if model_param_present or model_assignment:
```

Comment:
Should we use `and` instead of `or`?

Reply:
same as above, i was worried about model assignments without necessarily having the `model` param

Reply:
I see, can we just check the assignment to `self.model`?

Reply:
I think we should decrease the risk of false positives. False negatives are ok.

Reply:
Yes :)